cmd/package_managers.go

```diff
@@ -138,3 +206,134 @@ func fetchGitRepositoryFromRubyGems(packageName string, manager packageManagerCl
+	err = xml.NewDecoder(respPackageSpec.Body).Decode(packageSpecResults)
+
+	if err != nil || packageSpecResults.Metadata == (nuspecMetadata{}) {
+		return "", sce.WithMessage(sce.ErrScorecardInternal,
+			fmt.Sprintf("failed to parse nuget nuspec xml: %v", err))
+	}
+
+	if packageSpecResults.Metadata.Repository != (nuspecRepository{}) &&
+		len(strings.TrimSpace(packageSpecResults.Metadata.Repository.URL)) != 0 {
+		return packageSpecResults.Metadata.Repository.URL, nil
+	}
+	if len(strings.TrimSpace(packageSpecResults.Metadata.ProjectURL)) != 0 &&
+		isSupportedProjectURL(packageSpecResults.Metadata.ProjectURL) {
+		return packageSpecResults.Metadata.ProjectURL, nil
+	}
+	return "", sce.WithMessage(sce.ErrScorecardInternal,
+		fmt.Sprintf("source repo is not defined for nuget package %v", packageName))
+}
+
+func isSupportedProjectURL(projectURL string) bool {
+	pattern := `^(?:https?://)?(?:www\.)?(?:github|gitlab)\.com/([A-Za-z0-9_\.-]+)/([A-Za-z0-9_\.-]+)$`
+	regex := regexp.MustCompile(pattern)
+	return regex.MatchString(projectURL)
+}
+
+// Gets the latest listed nuget version of a package, based on the protocol defined at
+// https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource#enumerate-package-versions
+func getLatestListedVersion(baseURL, packageName string, manager packageManagerClient) (string, error) {
+	resPackageRegistrationIndex, err := manager.Get(baseURL+"%s/index.json", packageName)
+	if err != nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal,
+			fmt.Sprintf("failed to get nuget package registration index json: %v", err))
+	}
+	defer resPackageRegistrationIndex.Body.Close()
+	packageRegistrationCatalogRoot := &nugetPackageRegistrationCatalogRoot{}
+	err = json.NewDecoder(resPackageRegistrationIndex.Body).Decode(packageRegistrationCatalogRoot)
+	if err != nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal,
+			fmt.Sprintf("failed to parse package registration index json: %v", err))
+	}
+	return getLatestListedVersionFromPackageRegistrationPages(packageRegistrationCatalogRoot.Pages, manager)
+}
+
+func getLatestListedVersionFromPackageRegistrationPages(pages []nugetPackageRegistrationCatalogPage,
+	manager packageManagerClient,
+) (string, error) {
+	for pageIndex := len(pages) - 1; pageIndex >= 0; pageIndex-- {
+		page := pages[pageIndex]
+		if page.Packages == nil {
+			respPage, err := manager.GetURI(page.ID)
+			if err != nil {
+				return "", sce.WithMessage(sce.ErrScorecardInternal,
+					fmt.Sprintf("failed to get nuget package registration page: %v", err))
+			}
+			defer respPage.Body.Close()
+			err = json.NewDecoder(respPage.Body).Decode(&page)
+
+			if err != nil {
+				return "", sce.WithMessage(sce.ErrScorecardInternal,
+					fmt.Sprintf("failed to parse nuget package registration page: %v", err))
+			}
+		}
+		for packageIndex := len(page.Packages) - 1; packageIndex >= 0; packageIndex-- {
+			semVersion, err := semver.NewVersion(page.Packages[packageIndex].Entry.Version)
```

Comment:
I don't know the behavior of `semver` on a package version with 4 digits. NuGet diverged slightly from SemVer to allow compat with a popular .NET versioning scheme (System.Version) that has 4 digits. 

Example: Google.Apis.Storage.v1 has this version `1.60.0.2981`.

It's a valid version to NuGet but not a valid string per the strict SemVer spec. Here's an interesting doc we have about how NuGet and SemVer diverge in a couple nuanced places:
https://learn.microsoft.com/en-us/nuget/concepts/package-versioning#where-nugetversion-diverges-from-semantic-versioning

Reply:
this one actually broke the golang semver. switched to string splitting.